# Tutorial: Python for Computational Molecular Evolutionary Tasks

In this file, I provided Python scripts for working with biological sequences and phylogenetic trees. 
This tutorial assumes you are familiar with basic bioinformatic tasks and their extension to computational evolutionary models.

## Requirments
For this tutorial, you should already have [Python 3.9 or higher](https://realpython.com/installing-python/), [jupyter notebook](https://jupyter.org/install) along with the following libraries:

```numpy, ete3, biopython, seaborn, matplotlib and pandas```

### Installation

To install the package you can simply download the repository and run the following command in the root directory.

Install the dependencies using this command:

```console
pip3 install -r requirements.txt
```


## Lesson 1: Basics Sequence tasks

Shows how to load sequence data and perform simple data structure tasks.

## Lesson 2: Basics tree tasks

Shows how to load phylogenetic tree and perform simple data structure tasks.

## Lesson 3: Visualizing reconstructed ancestors using ARPIP

Visualizing posterior probability profile, indel points and etc.

## Lesson 4: Indel metrics

Measuring the dynamic of gaps and indels. 

## Lesson Bonus 

Converting existing exchangalbility matrices to generator 'Q' matrix for WAG and LG models.

## Contributing

If you want to contribute to this repo, submit a pull request!

## Author

Gholam-Hossein Jowkar [E-mail](jowk@zhaw.ch)
